### PR TITLE
[TE]  detection health status

### PR DESCRIPTION
**Pull request by jihaozh**
- The endpoint to return the detection health status in a certain time range.
- The detection health status builder

### CODE REVIEW

File: thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/DetectionHealth.java [hunk 1]
```diff
@@ -0,0 +1,379 @@
...
+          intervals.stream().map(interval -> interval.getEndMillis() - interval.getStartMillis()).reduce(0L, Long::sum);
+
+      AnomalyCoverageStatus coverageStatus = new AnomalyCoverageStatus();
+      coverageStatus.anomalyCoverageRatio = (double) totalAnomalyCoverage / (this.endTime - this.startTime);
+      coverageStatus.healthStatus = classifyCoverageStatus(coverageStatus.anomalyCoverageRatio);
+      return coverageStatus;
+    }
+
+    private static HealthStatus classifyCoverageStatus(double anomalyCoverageRatio) {
+      if (anomalyCoverageRatio > 0.85 || anomalyCoverageRatio < 0.01) {
```
Comment (akshayrai) [inline thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/DetectionHealth.java]: Define constant variables for thresholds. Here and elsewhere.

File: thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/DetectionResource.java [hunk 2]
```diff
@@ -562,4 +564,19 @@ public Response getPredictedBaseline(
...
 
+
+  @GET
+  @Path(value = "/health/{id}")
+  @ApiOperation("Get the detection health metrics and statuses for a detection config")
+  public Response getDetectionHealth(@PathParam("id") @ApiParam("detection config id") long id,
+      @ApiParam("Start time for the the health metric") @QueryParam("start") long start,
+      @ApiParam("End time for the the health metric") @QueryParam("end") long end,
+      @ApiParam("Max number of detection tasks returned") @QueryParam("limit") @DefaultValue("500") long limit) throws JsonProcessingException {
+    DetectionHealth health = new DetectionHealth.Builder(id, start, end).addRegressionStatus(this.evaluationDAO)
```
Comment (akshayrai) [inline thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/DetectionResource.java]: Add a log statement. Consider wrapping this in try-catch similar to YamlResource.

File: thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/DetectionHealth.java [hunk 3]
```diff
@@ -0,0 +1,379 @@
...
+        return HealthStatus.GOOD;
+      }
+      if (detectorHealthStatus.values().contains(HealthStatus.MODERATE)) {
+        return HealthStatus.MODERATE;
+      }
+      return HealthStatus.BAD;
+    }
+
+    private AnomalyCoverageStatus buildAnomalyCoverageStatus() {
+      List<MergedAnomalyResultDTO> anomalies = this.anomalyDAO.findByPredicate(
```
Comment (akshayrai) [inline thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/DetectionHealth.java]: You could enclose these builder methods in try-catch and throw a meaningful message.

File: thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/DetectionHealth.java [hunk 4]
```diff
@@ -0,0 +1,379 @@
...
+            end = anomaly.getEndTime();
+          }
+        }
+        intervals.add(new Interval(start, end));
+      }
+      long totalAnomalyCoverage =
+          intervals.stream().map(interval -> interval.getEndMillis() - interval.getStartMillis()).reduce(0L, Long::sum);
+
+      AnomalyCoverageStatus coverageStatus = new AnomalyCoverageStatus();
+      coverageStatus.anomalyCoverageRatio = (double) totalAnomalyCoverage / (this.endTime - this.startTime);
```
Comment (akshayrai) [inline thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/DetectionHealth.java]: Do we perform any validation on endTime and startTime? What if start is greater than end or if both are same!

File: thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/DetectionHealth.java [hunk 5]
```diff
@@ -0,0 +1,379 @@
+package org.apache.pinot.thirdeye.detection;
```
[the remark this answers is not in the document]
Comment (jihaozh) [reply thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/DetectionHealth.java]: Added

File: thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/DetectionHealth.java [hunk 6]
```diff
@@ -0,0 +1,379 @@
...
+
+  // the detection task status for a detection config
+  @JsonProperty
+  private DetectionTaskStatus detectionTaskStatus;
+
+  public enum HealthStatus {
+    GOOD, MODERATE, BAD
+  }
+
+  public static class RegressionStatus {
```
[the remark this answers is not in the document]
Comment (jihaozh) [reply thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/DetectionHealth.java]: Moved the classes

File: thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/DetectionHealth.java (hunk 1 above)
Comment (jihaozh) [reply thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/DetectionHealth.java]: Good point

File: thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/DetectionResource.java (hunk 2 above)
Comment (jihaozh) [reply thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/DetectionResource.java]: Added the try-catch

File: thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/DetectionHealth.java (hunk 3 above)
Comment (jihaozh) [reply thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/DetectionHealth.java]: Added the try-catch in the outer level

File: thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/DetectionHealth.java (hunk 4 above)
Comment (jihaozh) [reply thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/DetectionHealth.java]: We have this validation in the constructor

**Pull request merged**